### PR TITLE
Feature/voting incentives

### PR DESCRIPTION
**Pull request by razww**
Added the Bribe (Voting Incentives) module, allowing users to incentivize veLista holders to vote on the distribution of LP emissions

Use this cmd to run unit tests:
```
forge test --match-contract VotingIncentiveTest
```
image

### CODE REVIEW

File: contracts/dao/VotingIncentive.sol [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+    __AccessControl_init();
+    __Pausable_init();
+
+    _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+
+    vault = IVault(_vault);
+    emissionVoting = IEmissionVoting(_emissionVoting);
+    require(emissionVoting.hasRole(emissionVoting.ADMIN_VOTER(), _adminVoter), "Invalid adminVoter");
+    adminVoter = _adminVoter;
+    grantRole(DEFAULT_ADMIN_ROLE, _admin);
```
Comment (razww) [inline contracts/dao/VotingIncentive.sol:77]: grant pauser role in `initialize`

File: contracts/dao/VotingIncentive.sol [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+
+    startWeek = vault.getWeek(block.timestamp);
+  }
+
+  /**
+   * @dev Whiltelist an asset for voting incentives; zero address to whitelist native token (BNB)
+   * @param _asset address of the asset
+   * @param _whitelist true/false; true to whitelist, false to remove from whitelist
+   */
+  function whitelistAsset(address _asset, bool _whitelist) external onlyRole(DEFAULT_ADMIN_ROLE) {
```
Comment (razww) [inline contracts/dao/VotingIncentive.sol:87]: change permission to manager

File: contracts/dao/VotingIncentive.sol [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+   * @param _startWeek start week
+   * @param _endWeek end week
+   */
+  function addIncentivesBnb(uint16 _distributorId, uint16 _startWeek, uint16 _endWeek) external payable whenNotPaused {
+    require(assetWhitelist[address(0)], "Bnb not whitelisted");
+    require(msg.value > 0, "Invalid amount");
+    require(_distributorId > 0 && _distributorId <= vault.distributorId(), "Invalid distributorId");
+    require(
+      _startWeek >= (vault.getWeek(block.timestamp) + 1) && _startWeek <= _endWeek,
+      "Only current or future weeks"
```
Comment (razww) [inline contracts/dao/VotingIncentive.sol:105]: Only future weeks

File: contracts/dao/VotingIncentive.sol [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+    );
+
+    uint256 numberOfWeeks = uint256(_endWeek) - uint256(_startWeek) + 1;
+    uint256 weeklyAmount = msg.value / numberOfWeeks;
+
+    // Ensure input amount is divisible by number of weeks
+    require(weeklyAmount * numberOfWeeks == msg.value, "Input amount not divisible by number of weeks");
```
Comment (razww) [inline contracts/dao/VotingIncentive.sol:112]: remove this line

add emergencyWithdraw() method

File: contracts/dao/VotingIncentive.sol [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+    address _user,
+    uint16 _distributorId,
+    uint16 _week,
+    address _asset,
+    uint256 _adminWeight
+  ) internal view returns (uint256 _amount) {
+    uint256 poolWeight = emissionVoting.getDistributorWeeklyTotalWeight(_distributorId, _week);
+    uint256 usrWeight = getRawWeight(_user, _distributorId, _week);
+
+    require(poolWeight > _adminWeight, "Invalid pool weight");
```
Comment (razww) [inline contracts/dao/VotingIncentive.sol:218]: remove this line

File: contracts/dao/VotingIncentive.sol [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+
+    // If admin has voted, adjust user weight by removing admin weight from pool
+    uint256 factor = _adminWeight > 0 ? (1e18 * poolWeight) / (poolWeight - _adminWeight) : 1e18;
+
+    uint256 incentive = weeklyIncentives[_distributorId][_week][_asset];
+    _amount = (usrWeight * incentive * factor) / poolWeight / 1e18;
```
Comment (razww) [inline contracts/dao/VotingIncentive.sol:224]: simplify the calculation logic

File: contracts/dao/VotingIncentive.sol [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+
+    emit IncentiveAdded(_distributorId, _startWeek, _endWeek, _asset, actualAmount);
+  }
+
+  /**
+   * @dev Claim all incentives for a distributor for a week
+   * @param _distributorId id of the distributor
+   * @param _week week number
+   */
+  function batchClaim(uint16 _distributorId, uint16 _week, address[] memory _assets) external whenNotPaused {
```
Comment (razww) [inline contracts/dao/VotingIncentive.sol:167]: change `_distributorId` to array

Comment (qingyang-lista) [inline contracts/dao/VotingIncentive.sol:167]: we can make the batch methods like this
input:
`[{
  week
  distributeId,
  assets: [addres1, address2]
},
...]`

return:
`[{
  week,
  distributedId,
  assets: [{
    address1,
    amount1
  },{
    address2,
    amount2
  }]
},
....]`

File: contracts/dao/VotingIncentive.sol (hunk 1 above)
Comment (razww) [reply contracts/dao/VotingIncentive.sol]: done in 4051215ce3c8d5303d5a4438f78fc679a51afb93

File: contracts/dao/VotingIncentive.sol (hunk 2 above)
Comment (razww) [reply contracts/dao/VotingIncentive.sol]: done in 4051215ce3c8d5303d5a4438f78fc679a51afb93

File: contracts/dao/VotingIncentive.sol [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+  mapping(address => bool) public assetWhitelist;
+
+  // user -> distributor -> week -> asset -> claimed
+  mapping(address => mapping(uint16 => mapping(uint16 => mapping(address => bool)))) public claimedIncentives;
+
+  // distributorId -> week -> asset -> amount
+  mapping(uint16 => mapping(uint16 => mapping(address => uint256))) public weeklyIncentives;
+
+  // @dev responsible to halt the contract
+  bytes32 public constant PAUSER = keccak256("PAUSER");
```
[the remark this answers is not in the document]
Comment (razww) [reply contracts/dao/VotingIncentive.sol:43]: updated in 4051215ce3c8d5303d5a4438f78fc679a51afb93

File: contracts/dao/VotingIncentive.sol (hunk 3 above)
Comment (razww) [reply contracts/dao/VotingIncentive.sol]: done in 4051215ce3c8d5303d5a4438f78fc679a51afb93

File: contracts/dao/VotingIncentive.sol (hunk 5 above)
Comment (razww) [reply contracts/dao/VotingIncentive.sol]: removed in 4051215ce3c8d5303d5a4438f78fc679a51afb93

File: contracts/dao/VotingIncentive.sol (hunk 7 above)
Comment (razww) [reply contracts/dao/VotingIncentive.sol]: Updated in 4051215ce3c8d5303d5a4438f78fc679a51afb93

Added a new function `getClaimableAmount` to support the query ABI

**Review by qingyang-lista — APPROVED** [review]
LGTM

File: contracts/dao/VotingIncentive.sol (hunk 7 above)
Comment (razww) [reply contracts/dao/VotingIncentive.sol]: added an array of struct to group params

**Review by qingyang-lista — APPROVED** [review]
LGTM

**Pull request merged**